In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv("Sugarcane_Irrigation_Dataset_5000.csv")

print("Dataset loaded successfully!")
print("Shape:", df.shape)

df.head()

Dataset loaded successfully!
Shape: (5000, 17)


,rainfall_mm,avg_temperature_celsius,sunlight_hours_per_day,nitrogen_n_kg_per_acre,phosphorus_p_kg_per_acre,potassium_k_kg_per_acre,soil_type,crop_duration_days,irrigation_frequency_per_month,pest_control_applied,seed_variety,yield_quintal_per_acre,soil_ph,relative_humidity_percent,wind_speed_kmph,soil_moisture_percent,water_requirement_liters_per_day
0,1734.13,33.61,8.95,214.67,76.90,94.82,Alluvial,400,3,1.00,Co86032,520.17,6.82,88.9,10.1,36.1,15.9
1,1286.61,31.25,7.89,163.85,61.71,80.61,Alluvial,374,3,0.95,CoM0265,463.89,6.99,82.2,8.1,28.1,18.2
2,1323.73,29.54,7.52,133.31,48.41,65.28,Red Soil,352,2,0.99,Co0238,343.14,6.50,77.9,8.6,35.0,18.5
3,1459.15,32.62,8.55,190.85,69.67,82.79,Clay Loam,384,4,1.00,Co0238,508.21,6.83,87.5,10.7,33.0,19.6
4,1114.87,30.31,7.92,148.93,57.00,71.42,Clay Loam,359,4,1.00,CoM0265,433.10,7.10,79.0,20.1,25.8,18.6


In [ ]:
df = pd.read_csv("Sugarcane_Irrigation_Dataset_5000.csv")

print("Dataset loaded successfully!")
print("Shape:", df.shape)

df.head()

Dataset loaded successfully!
Shape: (5000, 17)


,rainfall_mm,avg_temperature_celsius,sunlight_hours_per_day,nitrogen_n_kg_per_acre,phosphorus_p_kg_per_acre,potassium_k_kg_per_acre,soil_type,crop_duration_days,irrigation_frequency_per_month,pest_control_applied,seed_variety,yield_quintal_per_acre,soil_ph,relative_humidity_percent,wind_speed_kmph,soil_moisture_percent,water_requirement_liters_per_day
0,1734.13,33.61,8.95,214.67,76.90,94.82,Alluvial,400,3,1.00,Co86032,520.17,6.82,88.9,10.1,36.1,15.9
1,1286.61,31.25,7.89,163.85,61.71,80.61,Alluvial,374,3,0.95,CoM0265,463.89,6.99,82.2,8.1,28.1,18.2
2,1323.73,29.54,7.52,133.31,48.41,65.28,Red Soil,352,2,0.99,Co0238,343.14,6.50,77.9,8.6,35.0,18.5
3,1459.15,32.62,8.55,190.85,69.67,82.79,Clay Loam,384,4,1.00,Co0238,508.21,6.83,87.5,10.7,33.0,19.6
4,1114.87,30.31,7.92,148.93,57.00,71.42,Clay Loam,359,4,1.00,CoM0265,433.10,7.10,79.0,20.1,25.8,18.6


In [ ]:
attributes = [
    "rainfall_mm",
    "avg_temperature_celsius",
    "yield_quintal_per_acre"
]

selected_data = df[attributes]

selected_data.head()

,rainfall_mm,avg_temperature_celsius,yield_quintal_per_acre
0,1734.13,33.61,520.17
1,1286.61,31.25,463.89
2,1323.73,29.54,343.14
3,1459.15,32.62,508.21
4,1114.87,30.31,433.10


Rainfall and temperature represent important environmental factors, while yield represents sugarcane productivity.

The frequency distributions divide rainfall, temperature and yield into five ranges and show the number of records in each range. This helps identify the ranges in which most sugarcane fields fall

**FREQUENCY DISTRIBUTION USING STRUGES' RULE**



In [ ]:
import math

n = len(df)

k = round(1 + 3.322 * math.log10(n))

print("Number of observations:", n)
print("Number of class intervals:", k)

Number of observations: 5000
Number of class intervals: 13


We calculated the number of classes and divided it into 13 class intervals

we use struges' rule to divide it into intervals as the dataset as around 5000 entries. The rule helps us identify in how many intervals should we divde it

**RAINFALL**

In [ ]:
rainfall_frequency = pd.cut(
    df["rainfall_mm"],
    bins=k
).value_counts().sort_index()

print("Frequency Distribution of Rainfall:")
print(rainfall_frequency)

Frequency Distribution of Rainfall:
rainfall_mm
(798.83, 890.0]     638
(890.0, 980.0]      517
(980.0, 1070.0]     308
(1070.0, 1160.0]    323
(1160.0, 1250.0]    294
(1250.0, 1340.0]    421
(1340.0, 1430.0]    320
(1430.0, 1520.0]    333
(1520.0, 1610.0]    267
(1610.0, 1700.0]    415
(1700.0, 1790.0]    468
(1790.0, 1880.0]    361
(1880.0, 1970.0]    335
Name: count, dtype: int64


**USING IN BUILT FUNCTIONS**

In [ ]:
rainfall_mean = df["rainfall_mm"].mean()

print("Mean Rainfall:", rainfall_mean, "mm")
rainfall_median = df["rainfall_mm"].median()

print("Median Rainfall:", rainfall_median, "mm")
rainfall_mode = df["rainfall_mm"].mode()

print("Mode of Rainfall:")
print(rainfall_mode)

Mean Rainfall: 1350.8812300000002 mm
Median Rainfall: 1339.655 mm
Mode of Rainfall:
0    800.0
Name: rainfall_mm, dtype: float64


The mean is the average value of each attribute .The middlemost value for each attribute is the median
The mode is the value that occurs most frequently.


**TEMPERATURE**

In [ ]:
temperature_frequency = pd.cut(
    df["avg_temperature_celsius"],
    bins=k
).value_counts().sort_index()

print("Frequency Distribution of Average Temperature:")
print(temperature_frequency)

Frequency Distribution of Average Temperature:
avg_temperature_celsius
(26.491, 27.169]    661
(27.169, 27.838]    346
(27.838, 28.508]    245
(28.508, 29.177]    240
(29.177, 29.846]    256
(29.846, 30.515]    300
(30.515, 31.185]    259
(31.185, 31.854]    428
(31.854, 32.523]    464
(32.523, 33.192]    593
(33.192, 33.862]    495
(33.862, 34.531]    458
(34.531, 35.2]      255
Name: count, dtype: int64


**USING INBUILT CODE**

In [ ]:
temperature_mean = df["avg_temperature_celsius"].mean()

print("Mean Temperature:", temperature_mean, "°C")
temperature_median = df["avg_temperature_celsius"].median()

print("Median Temperature:", temperature_median, "°C")
temperature_mode = df["avg_temperature_celsius"].mode()

print("Mode of Temperature:")
print(temperature_mode)


Mean Temperature: 30.93791 °C
Median Temperature: 31.49 °C
Mode of Temperature:
0    26.5
Name: avg_temperature_celsius, dtype: float64


**YIELD**

In [ ]:
yield_frequency = pd.cut(
    df["yield_quintal_per_acre"],
    bins=k
).value_counts().sort_index()

print("Frequency Distribution of Sugarcane Yield:")
print(yield_frequency)

Frequency Distribution of Sugarcane Yield:
yield_quintal_per_acre
(188.008, 218.554]    384
(218.554, 248.708]    122
(248.708, 278.862]    297
(278.862, 309.015]    491
(309.015, 339.169]    507
(339.169, 369.323]    483
(369.323, 399.477]    163
(399.477, 429.631]    189
(429.631, 459.785]    272
(459.785, 489.938]    542
(489.938, 520.092]    638
(520.092, 550.246]    564
(550.246, 580.4]      348
Name: count, dtype: int64


**USING IN BUILT FUNCTIONS**

In [ ]:
yield_mean = df["yield_quintal_per_acre"].mean()

print("Mean Yield:", yield_mean, "quintal/acre")
yield_median = df["yield_quintal_per_acre"].median()

print("Median Yield:", yield_median, "quintal/acre")
yield_mode = df["yield_quintal_per_acre"].mode()

print("Mode of Yield:")
print(yield_mode)

Mean Yield: 400.432326 quintal/acre
Median Yield: 408.4 quintal/acre
Mode of Yield:
0    188.4
Name: yield_quintal_per_acre, dtype: float64


**USING MANUAL CODE**

In [ ]:
#  Mean
def calculate_mean(data):
    total = 0

    for value in data:
        total = total + value

    mean = total / len(data)

    return mean


#  Median
def calculate_median(data):
    sorted_data = sorted(data)

    n = len(sorted_data)

    if n % 2 == 0:
        middle1 = sorted_data[n // 2 - 1]
        middle2 = sorted_data[n // 2]

        median = (middle1 + middle2) / 2

    else:
        median = sorted_data[n // 2]

    return median


# Mode
def calculate_mode(data):
    frequency = {}

    # Count frequency of every value
    for value in data:

        if value in frequency:
            frequency[value] = frequency[value] + 1

        else:
            frequency[value] = 1

    # Find highest frequency
    highest_frequency = 0

    for value in frequency:

        if frequency[value] > highest_frequency:
            highest_frequency = frequency[value]

    # If every value occurs only once
    if highest_frequency == 1:
        return "No mode"

    # Find all modes
    modes = []

    for value in frequency:

        if frequency[value] == highest_frequency:
            modes.append(value)

    return modes
    rainfall = df["rainfall_mm"].tolist()

print("RAIN FALL")
print("------------------------------")

print("Mean:", calculate_mean(rainfall))
print("Median:", calculate_median(rainfall))
print("Mode:", calculate_mode(rainfall))
temperature = df["avg_temperature_celsius"].tolist()

print("AVERAGE TEMPERATURE")
print("------------------------------")

print("Mean:", calculate_mean(temperature))
print("Median:", calculate_median(temperature))
print("Mode:", calculate_mode(temperature))
yield_data = df["yield_quintal_per_acre"].tolist()

print("SUGARCANE YIELD")
print("------------------------------")

print("Mean:", calculate_mean(yield_data))
print("Median:", calculate_median(yield_data))
print("Mode:", calculate_mode(yield_data))

RAIN FALL
------------------------------
Mean: 1350.881230000004
Median: 1339.655
Mode: [800.0]
AVERAGE TEMPERATURE
------------------------------
Mean: 30.93791000000005
Median: 31.49
Mode: [26.5]
SUGARCANE YIELD
------------------------------
Mean: 400.43232599999857
Median: 408.4
Mode: [188.4]


The average rainfall across the 5,000 sugarcane records is 1350.88 mm, while the median is 1339.66 mm. The mode is 800 mm, indicating that this is the most frequently occurring exact rainfall value. The frequency distribution shows that the 799.50–890.85 mm interval has the highest frequency with 638 records

The average temperature of the sugarcane fields is 30.94 °C, with a median of 31.49 °C. The mode is 26.5 °C, which is the most frequently occurring exact temperature value. The 26.50–27.17 °C interval contains the highest number of records (661).

he average sugarcane yield is 400.43 quintal/acre, while the median yield is 408.4 quintal/acre. The most frequently occurring exact yield value is 188.4 quintal/acre. The 489.90–520.05 quintal/acre interval has the highest frequency with 638 records.

The mean gives us gives the overall average rainfall /temperature condition, which provides a baseline for estimating natural water availability and the general level of crop water demand.For yield, mean shows us whioch improved irrigation techniques can be compared

The dataset is reasonably well distributed, with rainfall being almost symmetric and temperature and yield showing only slight negative skewness

All this data helps us prepare for reliable irrigation prediction and analysis.

**GROUPED ATTRIBUTES**

In [ ]:
# Grouped Data Analysis using two attributes:
# soil_type and seed_variety
# Statistical measures: Frequency, Mean, Median and Mode

#  Mean
def calculate_mean(data):
    total = 0

    for value in data:
        total = total + value

    return total / len(data)


#  Median
def calculate_median(data):
    sorted_data = sorted(data)
    n = len(sorted_data)

    if n % 2 == 0:
        middle1 = sorted_data[n // 2 - 1]
        middle2 = sorted_data[n // 2]
        return (middle1 + middle2) / 2
    else:
        return sorted_data[n // 2]


#  Mode
def calculate_mode(data):
    frequency = {}

    for value in data:
        if value in frequency:
            frequency[value] = frequency[value] + 1
        else:
            frequency[value] = 1

    highest_frequency = 0

    for value in frequency:
        if frequency[value] > highest_frequency:
            highest_frequency = frequency[value]

    # No repeated value
    if highest_frequency == 1:
        return "No mode"

    modes = []

    for value in frequency:
        if frequency[value] == highest_frequency:
            modes.append(value)

    return modes


# Group dataset by soil type and seed variety
grouped_data = df.groupby(["soil_type", "seed_variety"])

# Calculate frequency, mean, median and mode
for (soil, seed), group in grouped_data:

    yield_data = group["yield_quintal_per_acre"].tolist()

    print("\n" + "=" * 60)
    print("Soil Type    :", soil)
    print("Seed Variety :", seed)
    print("Frequency    :", len(yield_data))
    print("Mean Yield   :", calculate_mean(yield_data))
    print("Median Yield :", calculate_median(yield_data))
    print("Mode Yield   :", calculate_mode(yield_data))


Soil Type    : Alluvial
Seed Variety : Co0238
Frequency    : 171
Mean Yield   : 479.4921052631578
Median Yield : 479.3
Mode Yield   : [484.06, 488.82, 485.01]

Soil Type    : Alluvial
Seed Variety : Co86032
Frequency    : 789
Mean Yield   : 492.5676299112806
Median Yield : 529.62
Mode Yield   : [580.4]

Soil Type    : Alluvial
Seed Variety : CoM0265
Frequency    : 811
Mean Yield   : 483.1438964241679
Median Yield : 471.9
Mode Yield   : [566.19, 451.08, 545.84, 459.4, 395.98, 540.0, 557.16, 556.11, 391.77, 461.85, 475.26, 566.22, 456.13, 558.94, 472.84, 444.03, 388.0, 444.02, 466.94, 542.56, 554.51]

Soil Type    : Clay Loam
Seed Variety : Co0238
Frequency    : 531
Mean Yield   : 461.29994350282465
Median Yield : 500.88
Mode Yield   : [510.28, 357.18, 359.1, 356.43, 524.41, 507.68, 509.5, 498.94, 495.55, 495.33, 501.93, 361.33, 519.79, 494.28, 530.65, 526.59, 514.99, 511.86, 355.97]

Soil Type    : Clay Loam
Seed Variety : Co86032
Frequency    : 148
Mean Yield   : 493.97290540540564
Me

The dataset was grouped using two categorical attributes, soil type and seed variety, to analyze sugarcane yield under different cultivation combinations. Frequency, mean, median, and mode were calculated manually for each group. This helps compare productivity across different soil and seed-variety conditions and identify combinations associated with higher or more typical sugarcane yield.

CONVERTING A DISCRETE ATTRIBUTE INTO CONTINOUS

In [ ]:
import numpy as np


# Step 1: Convert discrete attribute to continuous


np.random.seed(42)

continuous_irrigation = (
    df["irrigation_frequency_per_month"].astype(float)
    + np.random.uniform(-0.49, 0.49, size=len(df))
)

df["continuous_irrigation_frequency"] = continuous_irrigation


# Step 2: Mean


def calculate_mean(data):
    total = 0

    for value in data:
        total = total + value

    return total / len(data)


# Step 3: Median

def calculate_median(data):
    sorted_data = sorted(data)
    n = len(sorted_data)

    if n % 2 == 0:
        middle1 = sorted_data[n // 2 - 1]
        middle2 = sorted_data[n // 2]

        return (middle1 + middle2) / 2

    else:
        return sorted_data[n // 2]


# Step 4: Mode


def calculate_mode(data):
    frequency = {}

    for value in data:

        if value in frequency:
            frequency[value] += 1

        else:
            frequency[value] = 1

    highest_frequency = 0

    for value in frequency:

        if frequency[value] > highest_frequency:
            highest_frequency = frequency[value]

    if highest_frequency == 1:
        return "No mode"

    modes = []

    for value in frequency:

        if frequency[value] == highest_frequency:
            modes.append(value)

    return modes


# Step 5: Apply calculations


continuous_data = df[
    "continuous_irrigation_frequency"
].tolist()

print("Continuous Irrigation Frequency")
print("=" * 45)

print("Mean   :", calculate_mean(continuous_data))
print("Median :", calculate_median(continuous_data))
print("Mode   :", calculate_mode(continuous_data))

Continuous Irrigation Frequency
Mean   : 3.282495351170622
Median : 3.1455149947392735
Mode   : No mode


The discrete irrigation_frequency_per_month attribute was converted into a continuous representation and analyzed .

**STANDARD DEVIATION AND VARIANCE**

In [ ]:
# Manual calculation
#  Variance
def calculate_variance(data):
    mean = calculate_mean(data)

    total = 0

    for value in data:
        total = total + (value - mean) ** 2

    # Sample variance
    variance = total / (len(data) - 1)

    return variance


#  Standard Deviation
def calculate_standard_deviation(data):
    variance = calculate_variance(data)

    return variance ** 0.5


# Apply to the selected attributes
attributes = [
    "rainfall_mm",
    "avg_temperature_celsius",
    "yield_quintal_per_acre"
]

for column in attributes:

    data = df[column].tolist()

    print("\nAttribute:", column)
    print("-" * 50)

    print(" Variance:",
          calculate_variance(data))

    print(" Standard Deviation:",
          calculate_standard_deviation(data))


Attribute: rainfall_mm
--------------------------------------------------
 Variance: 129134.53472597241
 Standard Deviation: 359.3529389416099

Attribute: avg_temperature_celsius
--------------------------------------------------
 Variance: 6.916895150930181
 Standard Deviation: 2.629999078123447

Attribute: yield_quintal_per_acre
--------------------------------------------------
 Variance: 13053.669232456168
 Standard Deviation: 114.25265525341706


In [ ]:
print("Variance and Standard Deviation using built in functions")
print("=" * 50)

for column in attributes:

    variance = df[column].var()
    standard_deviation = df[column].std()

    print("\nAttribute:", column)
    print("Variance:", variance)
    print("Standard Deviation:", standard_deviation)

Variance and Standard Deviation using built in functions

Attribute: rainfall_mm
Variance: 129134.53472597241
Standard Deviation: 359.3529389416099

Attribute: avg_temperature_celsius
Variance: 6.916895150930181
Standard Deviation: 2.629999078123447

Attribute: yield_quintal_per_acre
Variance: 13053.669232456168
Standard Deviation: 114.25265525341706


we calculated variance and standard deviation using both manual and built in functions. It helps us determine the consistency of the agricultural conditions from their average values . High rainfall SD shows rainfall conditions vary considerably across fields.
High temperature SD shoes different fields experience different temperature conditions.
High yield SD shows sugarcane productivity varies considerably.

The output shows that rainfall has substantial variability with a standard deviation of approximately 359.35 mm, while temperature is comparatively more consistent with a standard deviation of 2.63°C. Sugarcane yield has a standard deviation of approximately 114.25 quintal/acre, indicating a noticeable variation in crop productivity across the records.

**INTERQUARTILE RANGE**

In [ ]:
# IQR calculation

def calculate_median(data):
    data = sorted(data)
    n = len(data)

    if n % 2 == 0:
        return (data[n // 2 - 1] + data[n // 2]) / 2
    else:
        return data[n // 2]


def calculate_iqr(data):
    data = sorted(data)
    n = len(data)

    # Split data into lower and upper halves
    if n % 2 == 0:
        lower_half = data[:n // 2]
        upper_half = data[n // 2:]
    else:
        lower_half = data[:n // 2]
        upper_half = data[n // 2 + 1:]

    q1 = calculate_median(lower_half)
    q3 = calculate_median(upper_half)

    return q3 - q1


for column in attributes:

    data = df[column].tolist()

    print("\nAttribute:", column)
    print("Q1:", calculate_median(sorted(data)[:len(data)//2]))
    print("Q3:", calculate_median(sorted(data)[len(data)//2:]))
    print("IQR:", calculate_iqr(data))


Attribute: rainfall_mm
Q1: 1004.94
Q3: 1684.82
IQR: 679.8799999999999

Attribute: avg_temperature_celsius
Q1: 28.49
Q3: 33.144999999999996
IQR: 4.654999999999998

Attribute: yield_quintal_per_acre
Q1: 306.14
Q3: 503.83000000000004
IQR: 197.69000000000005


In [ ]:
print("IQR using built-in functions")
print("=" * 50)

for column in attributes:

    q1 = df[column].quantile(0.25)
    q3 = df[column].quantile(0.75)

    iqr = q3 - q1

    print("\nAttribute:", column)
    print("Q1:", q1)
    print("Q3:", q3)
    print("IQR:", iqr)

IQR using built-in functions

Attribute: rainfall_mm
Q1: 1005.03
Q3: 1684.815
IQR: 679.7850000000001

Attribute: avg_temperature_celsius
Q1: 28.49
Q3: 33.1425
IQR: 4.6525

Attribute: yield_quintal_per_acre
Q1: 306.145
Q3: 503.81
IQR: 197.66500000000002


The Interquartile Range is used to measure the spread of the middle 50% of observations. For rainfall, the middle 50% of values lie between 1005.03 mm and 1684.815 mm, giving an IQR of 679.785 mm, indicating considerable variation in typical rainfall conditions. For average temperature, the middle 50% lies between 28.49°C and 33.1425°C, with an IQR of 4.6525°C, showing relatively smaller variation in temperature. For sugarcane yield, the middle 50% lies between 306.145 and 503.81 quintal/acre, giving an IQR of 197.665 quintal/acre, indicating noticeable variation in typical crop productivity.

IQR is useful for the irrigation advisory system because it focuses on the central 50% of field conditions and is less affected by extreme values. The results show that rainfall has the greatest spread among the selected attributes, while temperature is comparatively stable. The substantial spread in yield indicates that sugarcane productivity varies considerably even within the central portion of the dataset, supporting the need for irrigation recommendations that account for differing field conditions.

In [ ]:
attributes = [
    "rainfall_mm",
    "avg_temperature_celsius",
    "yield_quintal_per_acre"
]

for column in attributes:

    q1 = df[column].quantile(0.25)
    q3 = df[column].quantile(0.75)

    iqr = q3 - q1

    # Mild outlier limits
    mild_lower = q1 - 1.5 * iqr
    mild_upper = q3 + 1.5 * iqr

    # Extreme outlier limits
    extreme_lower = q1 - 3 * iqr
    extreme_upper = q3 + 3 * iqr

    # Counts
    mild_outliers = (
        (df[column] < mild_lower) |
        (df[column] > mild_upper)
    ).sum()

    extreme_outliers = (
        (df[column] < extreme_lower) |
        (df[column] > extreme_upper)
    ).sum()

    print("\n" + "=" * 60)
    print("Attribute:", column)
    print("Q1:", q1)
    print("Q3:", q3)
    print("IQR:", iqr)

    print("\nMild Outlier Limits:")
    print("Lower:", mild_lower)
    print("Upper:", mild_upper)
    print("Number of Mild Outliers:", mild_outliers)

    print("\nExtreme Outlier Limits:")
    print("Lower:", extreme_lower)
    print("Upper:", extreme_upper)
    print("Number of Extreme Outliers:", extreme_outliers)


Attribute: rainfall_mm
Q1: 1005.03
Q3: 1684.815
IQR: 679.7850000000001

Mild Outlier Limits:
Lower: -14.64750000000015
Upper: 2704.4925000000003
Number of Mild Outliers: 0

Extreme Outlier Limits:
Lower: -1034.3250000000003
Upper: 3724.17
Number of Extreme Outliers: 0

Attribute: avg_temperature_celsius
Q1: 28.49
Q3: 33.1425
IQR: 4.6525

Mild Outlier Limits:
Lower: 21.511249999999997
Upper: 40.121249999999996
Number of Mild Outliers: 0

Extreme Outlier Limits:
Lower: 14.532499999999999
Upper: 47.099999999999994
Number of Extreme Outliers: 0

Attribute: yield_quintal_per_acre
Q1: 306.145
Q3: 503.81
IQR: 197.66500000000002

Mild Outlier Limits:
Lower: 9.647499999999923
Upper: 800.3075000000001
Number of Mild Outliers: 0

Extreme Outlier Limits:
Lower: -286.85000000000014
Upper: 1096.805
Number of Extreme Outliers: 0


The IQR method was used to detect mild and extreme outliers in rainfall, average temperature, and sugarcane yield. The analysis found no mild or extreme outliers in any of the three attributes. All 5,000 observations fall within the calculated outlier limits.

**GROUPED ANALYSIS OF VARIANCE, SD AND IQR**

In [ ]:
grouped_data = df.groupby(["soil_type", "seed_variety"])

for (soil, seed), group in grouped_data:

    yield_data = group["yield_quintal_per_acre"]

    variance = yield_data.var()
    standard_deviation = yield_data.std()

    q1 = yield_data.quantile(0.25)
    q3 = yield_data.quantile(0.75)
    iqr = q3 - q1

    print("\n" + "=" * 60)
    print("Soil Type    :", soil)
    print("Seed Variety :", seed)
    print("Frequency    :", len(yield_data))
    print("Variance     :", variance)
    print("Standard Deviation:", standard_deviation)
    print("IQR          :", iqr)


Soil Type    : Alluvial
Seed Variety : Co0238
Frequency    : 171
Variance     : 103.72583554179567
Standard Deviation: 10.184588138054266
IQR          : 13.425000000000011

Soil Type    : Alluvial
Seed Variety : Co86032
Frequency    : 789
Variance     : 8304.304557700423
Standard Deviation: 91.12795705874473
IQR          : 63.889999999999986

Soil Type    : Alluvial
Seed Variety : CoM0265
Frequency    : 811
Variance     : 3782.2572973917295
Standard Deviation: 61.50005932835943
IQR          : 103.32999999999993

Soil Type    : Clay Loam
Seed Variety : Co0238
Frequency    : 531
Variance     : 5246.9299771666165
Standard Deviation: 72.43569546271104
IQR          : 151.43499999999995

Soil Type    : Clay Loam
Seed Variety : Co86032
Frequency    : 148
Variance     : 91.98731735153521
Standard Deviation: 9.591001895085581
IQR          : 12.079999999999984

Soil Type    : Clay Loam
Seed Variety : CoM0265
Frequency    : 537
Variance     : 1325.830509958587
Standard Deviation: 36.411955590967

The grouped analysis shows that sugarcane yield variability differs substantially depending on the combination of soil type and seed variety. Some combinations, such as Alluvial–Co86032 and Clay Loam–Co0238, show large variation in productivity, whereas combinations such as Red Soil–CoM0265 and Sandy Loam–CoM0265 show much more consistent yields. This demonstrates that agricultural conditions are not uniform across fields. For an AI-based irrigation advisory system, this supports the need for condition-specific or plot-specific recommendations rather than applying one irrigation strategy to every sugarcane field